### DISCUSSION TITLE
`Astro.params` is greedy at rest params

### DISCUSSION THREAD

**Discussion by openscript** [withastro/roadmap#1023] [Proposal]
### Body

Huston! We have a problem with a greedy `Astro.params`.

## Background

From `getStaticPaths` I return the following `params` object:

`{ locale: undefined, notes: 'notes', note: 'tools/workstation' }`

The `Astro.params` receives:

`{ locale: 'notes', notes: 'tools', note: 'workstation' }`

For the page:

`src/pages/[...locale]/[notes]/[...note].astro`

It seems like when putting together `Astro.params` are treated greedy.

## Problem 

Astro's catch-all routes are greedy, meaning they will consume as much of the path as possible. In your case:

- `[...locale]` takes the first part of the path (`notes`).
- Since [notes] matches exactly one segment, tools is assigned to `[notes]`.
- `[...note]` gets whatever is left (`workstation`).

This is why locale becomes 'notes', even though you expect it to be undefined. Astro doesn't skip over parts of the URL—it tries to match everything based on the route structure you provided.

## Proposal

From other file based routers I know optional single segment params. Here would be handy to define such a param for `[locale?]` instead of using the rest param. Later on `Astro.param` could be put together as intended with the new information.

**Comment by Fryuni** [comment]
I am adding the info we discussed on Discord so others can find it.

> Astro's catch-all routes are greedy, meaning they will consume as much of the path as possible.

They are not, if they were the result would have been this:
```ts
{
  locale: 'notes/tools',
  notes: 'workstation',
  note: undefined,
}
```

You should not rely on any level of greediness or laziness of the matching.

> From other file based routers I know optional single segment params. Here would be handy to define such a param for `[locale?]` instead of using the rest param. Later on `Astro.param` could be put together as intended with the new information.

Optional parameters don't solve the ambiguity problem. `/[locale?]/[notes]/[...note]` would still be ambiguous for `/notes/tools/workstation`, allowing both the result you are seeing and the result you are expecting.

Having them is a valid proposal, but not a valid solution for case you presented.

---

To have the properties as you want, you can add them in your props as well instead of just the params:

```ts
export function getStaticPaths() {
  const params = { locale: undefined, notes: 'notes', note: 'tools/workstation' };
  return [{
    params, // Use with the route template to build the URL
    props: {
      params, // Passes as-is to the route props
      // other props you might already have
    },
  }];
}

const {locale, notes, note} = Astro.props.params;
```

When we reworked the routing logic on 4.2, there was a small discussion about whether multiple rest params should be allowed in the same route or prohibited to avoid this ambiguity problem. We ended up allowing users to express their intentions even if they would not get the values they passed in.
